### PR TITLE
WIP: Wandb logging

### PR DESCRIPTION
**Pull request by GainiK**
need to do the aggregation of results due to DDP

### CODE REVIEW

File: seg_lapa/config/trainer/trainer.yaml [hunk 1]
```diff
@@ -1,12 +1,12 @@
 # @package _group_
 name: trainer
-gpus: "0"
-overfit_batches: 0.0
+gpus: "0,1"
```
Comment (Shreeyak) [inline seg_lapa/config/trainer/trainer.yaml]:
```suggestion
gpus: 2
```
Lets keep this an int due to errors in pl with sending gpus ids that do not include gpu0 (I've made this change in my branch, not pushed yet).
See https://github.com/Shreeyak/pytorch-lightning-segmentation-template/pull/15

Comment (Shreeyak) [reply seg_lapa/config/trainer/trainer.yaml]: Very weird. The value shown in this diff is different from what is on master..
What I see as master in PR:
```
# @package _group_
name: trainer
gpus: "0"
overfit_batches: 0.0
```

Actual master branch
```

# @package _group_
name: trainer

gpus: 1  # Denotes the number of gpus to use. Set CUDA_VISIBLE_DEVICES env var to control which gpus are used.
overfit_batches: 0.0
```

Comment (Shreeyak) [reply seg_lapa/config/trainer/trainer.yaml]: The diff is against the point at which your branch started from master branch, instead of the latest master branch.

In the future, I think we should rebase the branch to master tip before PR, or merge master branch into the PR branch before submitting the PR.

Comment (Shreeyak) [reply seg_lapa/config/trainer/trainer.yaml]: I'll merge master into this PR before reviewing further.

File: seg_lapa/train.py [hunk 2]
```diff
@@ -27,20 +29,21 @@ def training_step(self, batch, batch_idx):
...
-
         return loss
 
     def validation_step(self, batch, batch_idx):
         inputs, labels = batch
         outputs = self.model(inputs)
         loss = self.cross_entropy_loss(outputs, labels)
+        batch_loss = loss/len(batch[0])
 
-        self.log("val_loss", loss, on_step=True, on_epoch=True, prog_bar=True, logger=True)
```
Comment (Shreeyak) [inline seg_lapa/train.py:43]: Wouldn't we still want some values to show up on the progress bar? passing `logger=False` will prevent values from being logged to wandb.

File: seg_lapa/train.py [hunk 3]
```diff
@@ -27,20 +29,21 @@ def training_step(self, batch, batch_idx):
...
         # "train_loss" is a reserved keyword
-        self.log("loss", loss, on_step=True, on_epoch=True, prog_bar=True, logger=True)
-
         return loss
 
     def validation_step(self, batch, batch_idx):
         inputs, labels = batch
         outputs = self.model(inputs)
         loss = self.cross_entropy_loss(outputs, labels)
+        batch_loss = loss/len(batch[0])
```
Comment (Shreeyak) [inline seg_lapa/train.py]: Let's rename it to avg_batch_loss, just to reduce confusion against loss vs batch_loss
```suggestion
        avg_batch_loss = loss/len(batch[0])
```

File: seg_lapa/train.py [hunk 4]
```diff
@@ -27,20 +29,21 @@ def training_step(self, batch, batch_idx):
...
 
-        self.log("val_loss", loss, on_step=True, on_epoch=True, prog_bar=True, logger=True)
+        wandb.log({"Val/BatchWise Loss": batch_loss})
```
Comment (Shreeyak) [inline seg_lapa/train.py:50]:
```suggestion
        wandb.log({"Val/BatchWise Loss": avg_batch_loss})
```
Same as above. The `"Val/BatchWise Loss"` name can remain as is

File: seg_lapa/train.py [hunk 5]
```diff
@@ -50,13 +53,21 @@ def test_step(self, batch, batch_idx):
         inputs, labels = batch
         outputs = self.model(inputs)
         loss = self.cross_entropy_loss(outputs, labels)
-
-        self.log("test_loss", loss, on_step=True, on_epoch=True, prog_bar=True, logger=True)
+        batch_loss = loss / len(batch[0])
+        wandb.log({"Test/Loss": batch_loss})
```
Comment (Shreeyak) [inline seg_lapa/train.py:61]: Same as prev 2 comments. Let's keep self.log, rename to avg_batch_loss.

File: seg_lapa/train.py [hunk 6]
```diff
@@ -27,20 +29,21 @@ def training_step(self, batch, batch_idx):
...
         loss = self.cross_entropy_loss(outputs, labels)
+        batch_loss = loss / len(batch[0])
+        wandb.log({"Train/BatchWise Loss": batch_loss})
 
         # to aggregate epoch metrics use self.log or a metric. self.log logs metrics for each training_step.
         # It also logs the average across the epoch, to the progress bar and logger
         # "train_loss" is a reserved keyword
-        self.log("loss", loss, on_step=True, on_epoch=True, prog_bar=True, logger=True)
-
         return loss
```
Comment (Shreeyak) [inline seg_lapa/train.py:41]: Why do we return single variable here, but dict in validation_step and test_step?
Which method is better?

File: seg_lapa/train.py [hunk 7]
```diff
@@ -74,6 +85,9 @@ def main(cfg: DictConfig):
     trainer = config.trainer.get_trainer()
     dm = config.dataset.get_datamodule()
 
+    wandb.init(project="segmentation", entity="cleargrasp2", name=str(trainer.logger.version))
```
Comment (Shreeyak) [inline seg_lapa/train.py:104]: Let's add an entry to config for these wandb values.

1. Why are we passing `name=str(trainer.logger.version)`? This will result in multiple runs from diff users/machines having the same name (because we're getting the local pl run number).

Suggestion: Pass `name=None` be default, and wandb will generate random name in dashboard. However, if user provides a run_name, use that instead.

2. Generate the random run_id ourselves. If we're resuming a run, the user would provide the run_id of resuming run, so the graph continues from where it left off.
```
run_id = wandb.util.generate_id()
resume_run = "never"  # Alternate: "allow"
wandb.init(..., id=run_id, resume=resume_run)
```

We'll work on actually resuming runs later.

3. These values for wandb should be present in config. We should be able to overwrite the project, entity, name etc from the command line.

File: seg_lapa/train.py [hunk 8]
```diff
@@ -74,6 +85,9 @@ def main(cfg: DictConfig):
...
+    wandb.watch(model)
```
Comment (Shreeyak) [inline seg_lapa/train.py:105]: I would think `wandb.watch()` was meant for pytorch nn.module. But here we pass in a pl.LightningModule. There aren't any side-effects? Does it behave as expected?

File: seg_lapa/train.py (hunk 6 above)
Comment (Shreeyak) [reply seg_lapa/train.py:41]: Looks like we can return whatever we want. Including the prediction and label tensors for computing metrics at train_epoch_end. We'll rework this as required for metrics.

File: seg_lapa/train.py (hunk 2 above)
Comment (Shreeyak) [reply seg_lapa/train.py:43]: Will look into if this is required. Looks like we'll end up using pl's logger anyway.

File: seg_lapa/train.py (hunk 7 above)
Comment (Shreeyak) [reply seg_lapa/train.py:104]: I'll take this up and submit separate PR.

File: seg_lapa/train.py (hunk 8 above)
Comment (Shreeyak) [reply seg_lapa/train.py:105]: Didn't see anything useful in wandb.watch(). I'm removing this in my PR.

File: seg_lapa/train.py [hunk 9]
```diff
@@ -50,17 +57,37 @@ def test_step(self, batch, batch_idx):
...
 
         return {
             "test_loss": loss,
         }
 
+    def training_epoch_end(self, outputs):
+        losses = torch.stack([x['loss'] for x in outputs])
+        gather_t = self.aggregate(losses)
+        gather_t = torch.flatten(torch.stack(gather_t))
+        loss = gather_t.mean()
```
Comment (Shreeyak) [inline seg_lapa/train.py:71]: I believe we don't need to use dist. functions in <mode>_step_end methods. They should return a list of all the outputs from all the batches.
I'll verify this.

File: seg_lapa/train.py [hunk 10]
```diff
@@ -50,17 +57,37 @@ def test_step(self, batch, batch_idx):
...
+        wandb.log({"Train/Epoch Loss": loss})
+
+    def validation_epoch_end(self, outputs):
+        losses = torch.stack([x['val_loss'] for x in outputs])
+        gather_t = self.aggregate(losses)
+        gather_t = torch.flatten(torch.stack(gather_t))
+        loss = gather_t.mean()
```
Comment (Shreeyak) [inline seg_lapa/train.py:78]: Same as above. We shouldn't need dist. functions in _step_end methods. I'll verify.

File: seg_lapa/train.py (hunk 9 above)
Comment (GainiK) [reply seg_lapa/train.py:71]: if we have 2 GPUs, the function train_epoch_end will be invoked twice with the outputs argument of length batch_size

**Comment by Shreeyak** [comment]
Closing this in favor of #17 
`torch.distributed` functions should not be required.

1. #17 got losses to aggregate using `self.log()`.
2. For metrics, we'll try to use the PL Metrics API - subclass the Metrics class to implement custom metrics. See: https://pytorch-lightning.readthedocs.io/en/latest/metrics.html#implementing-a-metric

File: seg_lapa/train.py (hunk 9 above)
Comment (Shreeyak) [reply seg_lapa/train.py:71]: Ah, OK.
